rig/routing_table/minimise.py

```diff
@@ -0,0 +1,557 @@
+            bottom = pos  # Move up
+        else:  # pg > generality
+            top = pos  # Move down
+
+        # Compute a new position
+        pos = bottom + (top - bottom) // 2
+        pg = gg(routing_table[pos])
+
+    while (pos < len(routing_table) and
+           gg(routing_table[pos]) <= generality):
+        pos += 1
+
+    return pos
+
+
+class Merge(namedtuple("Merge", ["routing_table", "entries", "key", "mask",
+                                 "generality", "goodness",
+                                 "insertion_index"])):
+    """Represents a potential merge of routing table entries.
+    """
+    def __new__(cls, routing_table, entries=set()):
+        # Generate the new key and mask
+        ones = 0x00000000  # Wherever there is a 1 in *any* of the keys
+        key = 0xffffffff  # ... 1 in *all* of the keys
+        mask = 0xffffffff  # ... 1 in *all* of the masks
+
+        for i in entries:
+            # Get the entry
+            entry = routing_table[i]
+
+            # Update the values
+            ones |= entry.key
+            key &= entry.key
+            mask &= entry.mask
+
+        # Compute the new mask, key and generality
+        # NOTE: `~key` indicates wherever there is a `0` in any key.
+        # Consequently `ones ^ ~key` indicates any bits in which there are
+        # either only zeros or ones (or, the bits where there doesn't need to
+        # be an X in the new entry).
+        mask &= (ones ^ ~key)  # Add any new Xs
+        key &= mask
+        generality = get_generality(key, mask)
+        insertion_index = get_insertion_index(routing_table, generality)
+
+        # Compute the goodness of the merge
+        goodness = len(entries) - 1
+
+        return super(Merge, cls).__new__(
+            cls, routing_table, frozenset(entries), key, mask,
+            generality, goodness, insertion_index
+        )
+
+    def apply(self, aliases):
+        """Apply the merge to the routing table it is defined against and get a
+        new routing table and alias dictionary.
+
+        Returns
+        -------
+        [:py:class:`~rig.routing_table.RoutingTableEntry`, ...]
+            A new routing table which may be minimised further.
+        {(key, mask): {(key, mask), ...}}
+            A new dictionary mapping keys and masks from the new routing table
+            to sets of keys and masks that they represent.
+        """
+        # Generate a new empty table of the correct size
+        routing_entries = list(self.routing_table)
+        aliases = dict(aliases)
+
+        # Add the new entry
+        new_entry = RoutingTableEntry(
+            route=routing_entries[next(iter(self.entries))].route,
+            key=self.key, mask=self.mask
+        )
+
+        routing_entries.insert(self.insertion_index, new_entry)
+        aliases[(self.key, self.mask)] = our_aliases = set([])
+
+        # Remove all the old entries
+        offset = 0
+        for i in sorted(self.entries):
```

Comment:
As I understand this relies on the fact that all merged entries must have lower generalities and thus be higher-up the table than the entry we just inserted meaning the insertion hasn't broken the indexes of the entries we care about? This should probably be explicitly pointed out.